core/solver/cg.hpp

```diff
@@ -89,106 +77,59 @@ class Cg : public BasicLinOp<Cg<ValueType>>, public PreconditionedMethod {
         return system_matrix_;
     }
 
-    /**
-     * Gets the maximum number of iterations of the CG solver.
-     *
-     * @return  The maximum number of iterations.
-     */
-    int get_max_iters() const { return max_iters_; }
-
-    /**
-     * Gets the relative residual goal of the solver.
-     *
-     * @return  The relative residual goal.
-     */
-    remove_complex<value_type> get_rel_residual_goal() const
+    std::shared_ptr<const LinOp> get_preconditioner() const
     {
-        return rel_residual_goal_;
+        return preconditioner_;
     }
 
-private:
-    using BasicLinOp<Cg>::create;
-
-    explicit Cg(std::shared_ptr<const Executor> exec)
-        : BasicLinOp<Cg>(exec, 0, 0, 0)
-    {}
-
-    Cg(std::shared_ptr<const Executor> exec, int max_iters,
-       remove_complex<value_type> rel_residual_goal,
-       std::shared_ptr<const LinOp> system_matrix)
-        : BasicLinOp<Cg>(
-              exec, system_matrix->get_num_cols(),
-              system_matrix->get_num_rows(),
-              system_matrix->get_num_rows() * system_matrix->get_num_cols()),
-          system_matrix_(std::move(system_matrix)),
-          max_iters_(max_iters),
-          rel_residual_goal_(rel_residual_goal)
-    {}
-
-    std::shared_ptr<const LinOp> system_matrix_{};
-    int max_iters_{};
-    remove_complex<value_type> rel_residual_goal_{};
-};
-
-
-/**
- * The CgFactory class is derived from the LinOpFactory class and is used to
- * generate the CG solver.
- */
-template <typename ValueType = default_precision>
-class CgFactory : public LinOpFactory, public PreconditionedMethodFactory {
-public:
-    using value_type = ValueType;
-    /**
-     * Creates the CG solver.
-     *
-     * @param exec The executor on which the CG solver is to be created.
-     * @param max_iters  The maximum number of iterations to be pursued.
-     * @param rel_residual_goal  The relative residual required for
-     * convergence.
-     *
-     * @return The newly created CG solver.
-     */
-    static std::unique_ptr<CgFactory> create(
-        std::shared_ptr<const Executor> exec, int max_iters,
-        remove_complex<value_type> rel_residual_goal)
+    GKO_ENABLE_LIN_OP_FACTORY(Cg, parameters, Factory)
```

Comment:
From looking at the code of the apply function of the Cg solver, I expect to find a `parameters_` member somewhere. What's really not clear here due to this macro (maybe I'm wrong, I haven't checked the macro because I want to provide an external look) is that this seems to stealthily add a `parameters_` member, right?

The problem to me is that we won't be able to find a definition/reference to `parameters_` with most editors and this may complexify startup time for beginners. Is there a way to change that? 

Maybe changing `parameters` to `parameters_` directly in the macro is enough, at least the name is the same so we can find it with direct name matching?

Reply:
Yes, I know it is an issue, but I'm not sure how to solve it.

The problem is that the names defined by the macro should follow the [style guide](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#members), so for parameters with name `foo`, the private variable should be named `foo_`, and its getter `get_foo()`. With macros I can add extra characters to names, but I cannot remove existing ones, so there's no way for me to generate `get_foo()`  if I get `foo_` as the macro parameter.

An alternative is not to generate the getter and the variable at all, and instead have the user do it (the user will usually want to do that, unless the factory has 0 parameters, which will rarely happen). However, then there is a problem that the user needs to know what is the type of the variable / getter, and the type is generated by the macro. If we also have the user define the type, then there's no reason to have the macro in the first place, as we can use `EnableDefaultFactory` directly, with minimal additions.

As I see it, as soon as we go into code generation, we'll need to have some conventions set up, otherwise it becomes very difficult or even impossible. The downside is that it will be harder for the editors and IDEs to figure out the code, but the alternative is to write more boilerplate code. Note that these conventions are part of the documentation for the `ENABLE_LIN_OP_FACTORY` macro, so you do know that if you pass `foo` as the parameter, it will generate a type called `foo_type`, a private member called `foo_` and a getter called `get_foo()`.

Reply:
You're right I forgot the getter function, that indeed complexifies things quite a bit. I do understand that the macro states what's going on so you can figure it out if you look at the documentation there. 

Another solution would be to be redundant in the macro, and pass both `parameters_` and ` parameters` used for the variable name definition and the getter (other generic things)? That's error prone but we can find `parameters_` easily, there's a clear tradeoff anyway.

Reply:
How about the `parameters_type`? Do we also pass that? If we pass `parameters_`, then it's only logical to also pass `parameters_type` (it just happens that right now we don't use it anywhere so you cannot see it, but it does appear from nowhere as well). Then we have `GKO_ENABLE_LIN_OP_FACTORY(Cg, parameters_type, parameters_, get_parameters, Factory)`.
Quite verbose and redundant.

The same problem is in the `GKO_FACTORY_PARAMETER(name, default)` macro. There, the `with_name()` function appears from nowhere. Do we also have the user specify that?

How about the mixins? They provide functions that are hard to track as well, and you cannot do anything about it.

Basically the entire code generation feature imports symbols which the editors get confused with.

Reply:
Well yes, I agree with you this MR adds a lot of things in this regard. Code navigation and finding declaration is already a bit confusing for example with the macros and struct which creates `make_xxx_operation()` functions and the surrounding class (though we can find a jumping point with the `struct XXXOperation` at the top of the file so that's OK-ish). But if we add that everywhere we really will kill easy code navigation in Ginkgo, with the declaration/definition of functions or member attributes not easily found.

I personally favor code duplication if it's reasonable and allows to simplify navigation (something we may very well do daily, and helps newcomers). What I'm wondering is if there is a standard way to e.g. insert directly using documentation or comments some sort of bookmarks/links/... which could solve this problem.

Do you have any input on this topic @ginkgo-project/developers (it's best to read the whole thread though).

Reply:
Should we also move this into a separate issue? I feel this is in general about code generation, and not just about this macro.

Reply:
I agree. I will do that for both discussions.